File: app/infrastructure/archive.py
"""Utility functions for reading archived files from ZIP storage.

Snapshots and videos older than the zip threshold are compressed into
``data/archives/`` ZIP files. This module provides helpers to extract
files from those archives without coupling filesystem logic into
routers or services.
"""

import zipfile
from pathlib import Path

from app.core.config import settings


def read_snapshot_from_archive(archive_ref: str) -> bytes:
    """Extract a snapshot image from a ZIP archive given its reference.

    The reference format is ``{rel_zip_path}::{filename}``, e.g.
    ``snapshots/3/2025-06-15.zip::150322.jpg``.

    Args:
        archive_ref: Archive reference string stored in the snapshot row.

    Returns:
        The raw bytes of the extracted JPEG image.

    Raises:
        FileNotFoundError: When the archive ZIP or the file inside it
            does not exist.
    """
    zip_rel, filename = archive_ref.split("::", 1)
    zip_abs = settings.archives_dir / zip_rel
    if not zip_abs.exists():
        raise FileNotFoundError(f"Archive not found: {zip_abs}")
    with zipfile.ZipFile(zip_abs, "r") as zf:
        if filename not in zf.namelist():
            raise FileNotFoundError(f"{filename} not found in {zip_abs}")
        return zf.read(filename)


def read_video_from_archive(filename: str) -> bytes | None:
    """Extract a video MP4 from an archive ZIP by filename.

    Derives the archive path from the filename pattern:
    ``timelapse_{camera_id}_{date}_h{hour}.mp4`` or
    ``timelapse_{camera_id}_{date}.mp4``.

    Args:
        filename: The MP4 filename to look up in the archive.

    Returns:
        The raw bytes of the MP4 file, or None when the archive or
        file inside it is not found.
    """
    parts = Path(filename).stem.split("_")
    if len(parts) < 3:
        return None
    cam_id = parts[1]
    date_part = parts[2]
    zip_path = settings.archives_dir / "videos" / cam_id / f"{date_part}.zip"
    if not zip_path.exists():
        return None
    with zipfile.ZipFile(zip_path, "r") as zf:
        if filename not in zf.namelist():
            return None
        return zf.read(filename)

File: internal/infrastructure/archive/archive.go
// Package archive reads files back from the retention ZIP archives.
// Snapshots store a reference of the form
// "{rel_zip_path}::{filename}", e.g. "snapshots/3/2025-06-15.zip::150322.jpg".
package archive

import (
	"archive/zip"
	"bytes"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"

	"github.com/Lmex89/home-cameras/internal/config"
)

// Reader resolves archived files against the archives directory.
type Reader struct {
	archivesDir string
}

// New builds a reader bound to the config's archives directory.
func New(cfg config.Config) *Reader {
	return &Reader{archivesDir: cfg.ArchivesDir()}
}

// ReadSnapshot extracts a snapshot JPEG given its archive reference.
func (r *Reader) ReadSnapshot(archiveRef string) ([]byte, error) {
	zipRel, filename, ok := strings.Cut(archiveRef, "::")
	if !ok || zipRel == "" || filename == "" {
		return nil, fmt.Errorf("invalid archive reference %q", archiveRef)
	}
	zipAbs := filepath.Join(r.archivesDir, zipRel)
	if _, err := os.Stat(zipAbs); err != nil {
		return nil, fmt.Errorf("archive not found: %s", zipAbs)
	}
	return readZipEntry(zipAbs, filename)
}

// ReadVideo extracts an MP4 from the videos archive by filename.
// Derives the archive path from the filename pattern
// "timelapse_{camera_id}_{date}*.mp4".
func (r *Reader) ReadVideo(filename string) ([]byte, error) {
	parts := strings.Split(strings.TrimSuffix(filename, filepath.Ext(filename)), "_")
	if len(parts) < 3 {
		return nil, fmt.Errorf("unrecognized video filename %q", filename)
	}
	camID, datePart := parts[1], parts[2]
	zipAbs := filepath.Join(r.archivesDir, "videos", camID, datePart+".zip")
	if _, err := os.Stat(zipAbs); err != nil {
		return nil, fmt.Errorf("archive not found: %s", zipAbs)
	}
	return readZipEntry(zipAbs, filename)
}

func readZipEntry(zipPath, name string) ([]byte, error) {
	zr, err := zip.OpenReader(zipPath)
	if err != nil {
		return nil, err
	}
	defer zr.Close()
	for _, f := range zr.File {
		if f.Name != name {
			continue
		}
		rc, err := f.Open()
		if err != nil {
			return nil, err
		}
		defer rc.Close()
		var buf bytes.Buffer
		if _, err := io.Copy(&buf, rc); err != nil {
			return nil, err
		}
		return buf.Bytes(), nil
	}
	return nil, fmt.Errorf("%s not found in %s", name, zipPath)
}
